rust/kernel/error.rs

```diff
@@ -104,3 +104,20 @@ impl From<AllocError> for Error {
         Error::ENOMEM
     }
 }
+
+pub(crate) unsafe fn ptr_err_check<T>(ptr: *mut T) -> KernelResult<*mut T> {
```

Comment:
@fbq has a different approach to this in #109. We should discuss which one we'd rather have.

Reply:
Yep, I'm OK with either implementation. One thing I wonder is whether Rust compiler could inline a extern "C" function? If it could, then I think @TheSven73 implementation is better, because we don't need to worry about kernel modifying the errno range (although it's a very rare case).

Reply:
I had a look at the `arm32` assembly code generated for the kernel. It doesn't look like the `rust_helper_is_err()/ptr_err()` is being inlined:

```asm
00014024 <_RNvMs0_NtCsbDqzXfLQacH_6kernel6regmapNtB5_6Regmap27init_mmio_platform_resource>:
   14024:       e92d48f0        push    {r4, r5, r6, r7, fp, lr}
   14028:       e24dd0a8        sub     sp, sp, #168    ; 0xa8
   1402c:       e1a06000        mov     r6, r0
   14030:       e5900000        ldr     r0, [r0]
   14034:       e1a04002        mov     r4, r2
   14038:       ebfffffe        bl      0 <devm_platform_ioremap_resource>
   1403c:       e1a05000        mov     r5, r0
   14040:       ebfffffe        bl      0 <rust_helper_is_err>
```
I guess there's a trade-off between performance/efficiency (an extra function call) and safety/compatibility (kernel changing its definitions on us).

Is it ok if we discuss this in a separate PR, and defer the decision? We can keep #109 as the place where these trade-offs are discussed. If #109 gets merged first. I'll rebase to it. If the opposite happens, you can rebase to mine. In both cases, the main branch will contain the consensus implementation of `ptr_to_result()` after both PRs are merged.

Or I'm open to anything else you'd like to suggest.

Reply:
I think it makes more sense if we have a specific PR for this change. I will make one and use my implementation, feel free to put any thought on it.

Reply:
done #267

Reply:
In general, I think we want to go for re-using the C functions/macros where possible. Performance work should be based on actual numbers, which we can focus on later. And, anyway, we should be able to have cross-language LTO at some point, which will most likely inline those.